runtime/lib/ttmetal/runtime.cpp

```diff
@@ -123,50 +130,138 @@ void closeDevice(Device device) {
-      inputs;
-  inputs.reserve(inputHandles.size());
-  assert(inputHandles.size() == program->inputs()->size() &&
-         "Input size mismatch");
-  for (unsigned i = 0; i < inputHandles.size(); ++i) {
-    inputs.emplace_back(
-        program->inputs()->Get(i)->global_id(),
-        static_pointer_cast<::tt::tt_metal::Buffer>(inputHandles[i].handle));
-  }
-
-  std::vector<std::pair<std::uint32_t, std::shared_ptr<::tt::tt_metal::Buffer>>>
-      outputs;
-  outputs.reserve(outputHandles.size());
-  assert(outputHandles.size() == program->outputs()->size() &&
-         "Output size mismatch");
-  for (unsigned i = 0; i < outputHandles.size(); ++i) {
-    outputs.emplace_back(
-        program->outputs()->Get(i)->global_id(),
-        static_pointer_cast<::tt::tt_metal::Buffer>(outputHandles[i].handle));
-  }
-
   DeviceMesh &deviceMesh = deviceHandle.as<DeviceMesh>();
+  assert(deviceMesh.size() == 1 && "Only one device is supported for now");
   std::shared_ptr<Events> events = std::make_shared<Events>();
-  std::size_t cq_id = 0;
   assert(program->device_programs()->size() == deviceMesh.size() &&
          "Device programs size mismatch");
   for (std::size_t i = 0; i < program->device_programs()->size(); ++i) {
     ::tt::tt_metal::Device *device = deviceMesh[i];
     ::tt::target::metal::DeviceProgram const *deviceProgram =
         program->device_programs()->Get(i);
+    Events deviceEvents;
+
+    std::vector<InputBuffer> inputs;
+    inputs.reserve(inputHandles.size());
+    assert(inputHandles.size() == deviceProgram->inputs()->size() &&
+           "Input size mismatch");
+    for (unsigned i = 0; i < inputHandles.size(); ++i) {
+      ::tt::target::TensorRef const *tensorRef =
+          deviceProgram->inputs()->Get(i);
+      auto [buffer, event] =
+          prepareInput(device, inputHandles[i].as<MetalTensor>(),
+                       inputHandles[i].data.get(), tensorRef);
+      inputs.emplace_back(deviceProgram->inputs()->Get(i)->global_id(), buffer,
+                          event);
+    }
+
+    std::vector<OutputBuffer> outputs;
+    outputs.reserve(outputHandles.size());
+    assert(outputHandles.size() == deviceProgram->outputs()->size() &&
+           "Output size mismatch");
+    for (unsigned i = 0; i < outputHandles.size(); ++i) {
+      ::tt::target::TensorRef const *tensorRef =
+          deviceProgram->outputs()->Get(i);
+      std::shared_ptr<::tt::tt_metal::Buffer> buffer =
+          prepareOutput(device, &outputHandles[i].as<MetalTensor>(), tensorRef);
+      outputs.emplace_back(deviceProgram->outputs()->Get(i)->global_id(),
+                           buffer);
+    }
+
+    std::size_t cq_id = 0;
     for (::tt::target::metal::CommandQueue const *cq :
          *deviceProgram->command_queues()) {
-      events->push_back(
+      deviceEvents.push_back(
           executeCommandQueue(device, cq, cq_id, inputs, outputs));
       ++cq_id;
     }
+
+    Events copyEvents =
```

Comment:
Just to double check, idea here is that if device outputs need to be copied to host, we will ignore earlier events generated from the execution of programs, by just returning the newer events that represent the copying (reads) of outputs from device to host... seems to make sense.

Reply:
Yes, though we don't ignore the earlier events, they are waited on inside of `maybeCopyHostOutputs`.